WordPress/src/main/java/org/wordpress/android/AppInitializer.kt

```diff
@@ -0,0 +1,981 @@
+import org.wordpress.android.ui.notifications.services.NotificationsUpdateServiceStarter
+import org.wordpress.android.ui.notifications.utils.NotificationsUtils
+import org.wordpress.android.ui.posts.editor.ImageEditorFileUtils
+import org.wordpress.android.ui.posts.editor.ImageEditorInitializer.Companion.init
+import org.wordpress.android.ui.posts.editor.ImageEditorTracker
+import org.wordpress.android.ui.prefs.AppPrefs
+import org.wordpress.android.ui.reader.tracker.ReaderTracker
+import org.wordpress.android.ui.stats.refresh.lists.widget.WidgetUpdater.StatsWidgetUpdaters
+import org.wordpress.android.ui.stories.media.StoryMediaSaveUploadBridge
+import org.wordpress.android.ui.uploads.UploadService
+import org.wordpress.android.ui.uploads.UploadStarter
+import org.wordpress.android.util.AppLog
+import org.wordpress.android.util.AppLog.T.API
+import org.wordpress.android.util.AppLog.T.DB
+import org.wordpress.android.util.AppLog.T.MAIN
+import org.wordpress.android.util.AppLog.T.NOTIFS
+import org.wordpress.android.util.AppLog.T.READER
+import org.wordpress.android.util.AppLog.T.UTILS
+import org.wordpress.android.util.AppThemeUtils.Companion.setAppTheme
+import org.wordpress.android.util.BitmapLruCache
+import org.wordpress.android.util.DateTimeUtils
+import org.wordpress.android.util.EncryptedLogging
+import org.wordpress.android.util.FluxCUtils
+import org.wordpress.android.util.LocaleManager
+import org.wordpress.android.util.NetworkUtils
+import org.wordpress.android.util.PackageUtils
+import org.wordpress.android.util.ProfilingUtils
+import org.wordpress.android.util.QuickStartUtils.cancelQuickStartReminder
+import org.wordpress.android.util.RateLimitedTask
+import org.wordpress.android.util.SiteUtils
+import org.wordpress.android.util.VolleyUtils
+import org.wordpress.android.util.WPActivityUtils
+import org.wordpress.android.util.analytics.AnalyticsUtils
+import org.wordpress.android.util.config.AppConfig
+import org.wordpress.android.util.enqueuePeriodicUploadWorkRequestForAllSites
+import org.wordpress.android.util.experiments.ExPlat
+import org.wordpress.android.util.image.ImageManager
+import org.wordpress.android.widgets.AppRatingDialog
+import org.wordpress.android.workers.WordPressWorkersFactory
+import java.io.File
+import java.io.IOException
+import java.util.Date
+import java.util.HashMap
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Suppress("TooManyFunctions")
+@Singleton
+class AppInitializer @Inject constructor() : LifecycleObserver {
+    // Even if `mWebViewDataDirectorySuffixInitializer`, `mWellSqlInitializer` aren't used, they need to be initialized
+    // at application startup.
+    @Inject lateinit var mWebViewDataDirectorySuffixInitializer: WebViewDataDirectorySuffixInitializer
+
+    // Well sql is initialized with `WPWellSqlConfig` for debug, `WellSqlConfig` for release.
+    @Inject lateinit var mWellSqlInitializer: WellSqlInitializer
```

Comment:
np, depending on the comment here to ensure WellSql is initialized is not the best approach IMO, since you added a separate class for the initialization, I would suggest adding an explicit `init` function to it and calling it from the `AppInitializer#init` function, instead of depending on the `init` bloc.
This would also ensure that the behavior is correct even if Dagger's behavior changed from eager injecting to lazier injecting here 🙂.

Same remark for `WebViewDataDirectorySuffixInitializer`.

Reply:
I initialized them with `@Inject` annotation instead of `AppInitializer#init` function, because other `@Inject` initializations need them.

If I init `WellSql` in `AppInitializer#init`, `AccountStore` crashes because of uninitialized `WellSql`.

I want to initialize `WebViewDataDirectorySuffixInitializer` and `WellSqlInitializer` before everything else. I think doing it in `AppInitializer#init` is not a right way to do that but we can discuss alternative ways.

What is your main concern here? Does it seem weird to inject a class just for initializing them?

Reply:
> If I init WellSql in AppInitializer#init, AccountStore crashes because of uninitialized WellSql.

Yes, I forgot about this, so maybe calling `init` directly in the app's `onCreate` 🤔 

> What is your main concern here? Does it seem weird to inject a class just for initializing them?

It's nothing major, but just two small issues IMO:
1. It's not super clear, and that's why we depend on the comment to explain why it's needed.
2. It depends on Dagger's logic, if Dagger was to inject the dependencies lazily by default (like Koin for example), this wouldn't work, since the class won't be instantiated until it's injected.

Anyway, it's just a small point, so whatever you and @renanferrari agree on, for WCAndroid, we didn't create any class or WellSql's initialization, and instead it's done manually in the Application's `onCreate`.

Reply:
> Yes, I forgot about this, so maybe calling init directly in the app's onCreate 🤔

It's called in [WooCommerce](https://github.com/woocommerce/woocommerce-android/blob/24dd260a478e55c6905e99bdd1012052ec655592/WooCommerce/src/main/kotlin/com/woocommerce/android/WooCommerce.kt#L27) in WCAndroid. At first I thought doing the same, but then I thought keeping application class minimal could be a better approach. Otherwise, I'd have to handle `WellSql` initialization also in test application, like [this](https://github.com/woocommerce/woocommerce-android/blob/0a85b39af97b2232000ca33fbd2e22b4f59ecabf/WooCommerce/src/androidTest/kotlin/com/woocommerce/android/WooCommerceTest.kt#L27) made in WCAndroid.

What about initialize them in `AppInitializer`'s init and injecting `AccountStore` lazily?
https://gist.github.com/irfano/790096d16ed4b0c4ca1adc366102d191

Reply:
> What about initialize them in AppInitializer's init and injecting AccountStore lazily?

Yes, this another possibility, I like it 👍